client/executor/wasmtime/src/runtime.rs

```diff
@@ -232,10 +232,75 @@ directory = \"{cache_dir}\"
+		native_stack_max, ..
+	}) = semantics.deterministic_stack_limit
+	{
+		config
+			.max_wasm_stack(native_stack_max as usize)
+			.map_err(|e| WasmError::Other(format!("cannot set max wasm stack: {}", e)))?;
+	}
+
+	// Be clear and specific about the extensions we support. If an update brings new features
+	// they should be introduced here as well.
+	config.wasm_reference_types(false);
+	config.wasm_simd(false);
+	config.wasm_bulk_memory(false);
+	config.wasm_multi_value(false);
+	config.wasm_multi_memory(false);
+	config.wasm_module_linking(false);
+	config.wasm_threads(false);
+
+	Ok(config)
+}
+
+/// Knobs for deterministic stack height limiting.
+///
+/// The WebAssembly standard defines a call/value stack but it doesn't say anything about its
+/// size except that it has to be finite. The implementations are free to choose their own notion
+/// of limit: some may count the number of calls or values, others would rely on the host machine
+/// stack and trap on reaching a guard page.
+///
+/// This obviously is a source of non-determinism during execution. This feature can be used
+/// to instrument the code so that it will count the depth of execution in some deterministic
+/// way (the machine stack limit should be so high that the deterministic limit always triggers
+/// first).
+///
+/// The deterministic stack height limiting feature allows to intstrument the code so that it will
+/// count the number of items that may be on the stack. This counting will only act as an rough
+/// estimate of the actual stack limit in wasmtime. This is because wasmtime measures it's stack
+/// usage in bytes.
+///
+/// The actual number of bytes consumed by a function is not trivial to compute  without going through
+/// full compilation. Therefore, it's expected that `native_stack_max` is grealy overestimated and
+/// thus never reached in practice. The stack overflow check introduced by the instrumentation and
+/// that relies on the logical item count should be reached first.
+///
+/// See [here][stack_height] for more details of the instrumentation
+///
+/// [stack_height]: https://github.com/paritytech/wasm-utils/blob/d9432baf/src/stack_height/mod.rs#L1-L50
+pub struct DeterministicStackLimit {
+	/// A number of logical "values" that can be pushed on the wasm stack. A trap will be triggered
+	/// if exceeded.
+	///
+	/// A logical value is a local, an argument or a value pushed on operand stack.
+	pub logical_max: u32,
+	/// The maximum number of bytes for stack used by wasmtime JITed code.
+	///
+	/// It's not specified how much bytes will be consumed by a stack frame for a given wasm function
+	/// after translation into machine code. It is also not quite trivial.
+	///
+	/// Therefore, this number should be choosen conservatively. It must be so large so that it can
+	/// fit the [`logical_max`] logical values on the stack, according to the current instrumentation
+	/// algorithm.
+	///
+	/// This value cannot be 0.
+	pub native_stack_max: u32,
```

Comment:
naming nit: I find it hard to understand the meaning without looking at the docs

how about `values_max`/`pushes_max` and `bytes_max`?

Reply:
you mean as prefix?

Reply:
I think what is important to convey here is that there are two limits in the context: the limit imposed by the instrumentation which is measured in some logical items and the other limit is of the native stack imposed by wasmtime and measured in bytes. 

It is more important what is the underlying restriction of the mechanism. The fact that they have different units of measure is secondary IMO.

Hah, now you said that, maybe we should rename it to something super obscure, like `rub_the_bear_belly`, just to make people read the comments more 😆